src/functions_dims.jl

```diff
@@ -21,10 +21,15 @@ end
 function Base.permutedims(nda::NamedDimsArray{L}, perm) where {L}
     numerical_perm = dim(nda, perm)
     new_names = permute_dimnames(L, numerical_perm)
-
     return NamedDimsArray{new_names}(permutedims(parent(nda), numerical_perm))
 end
 
+function Base.PermutedDimsArray(nda::NamedDimsArray{L}, perm) where {L}
+    numerical_perm = dim(nda, perm)
+    new_names = permute_dimnames(L, numerical_perm)
+    return NamedDimsArray{new_names}(PermutedDimsArray(parent(nda), numerical_perm))
```

Comment:
Shouldn't this return a `PermutedDimsArray` (wrapping a NamedDimsArray)?

Wouldn't we just want to add a constructor that allows `perm` to contain `Symbol`s?

Since `PermuteDimsArray` (with Ints) already works here, don't we just want:
```julia
julia> nda = NamedDimsArray{(:w, :x, :y, :z)}(ones(10, 20, 30, 40));

julia> pda = PermutedDimsArray(nda, (1, 2, 4, 3));  # already works

julia> typeof(pda)
PermutedDimsArray{Float64,4,(1, 2, 4, 3),(1, 2, 4, 3),NamedDimsArray{(:w, :x, :y, :z),Float64,4,Array{Float64,4}}}

julia> size(pda)
(10, 20, 40, 30)

julia> pda == PermutedDimsArray(nda, (:w, :x, :z, :y));  # this is the method that should be added?
true
```

Reply:
I was hoping `PermutedDimsArray` could return something with names you can refer to. Although @oxinabox raises the issue that it's weird not to return the constructor's type. Perhaps another objection is that if you care about the order of indices, then you are moving to positional not named code anyway, and should unwrap.

Reply:
What if we keep `PermutedDimsArray` returning a `PermutedDimsArray`, but add some methods on `PermutedDimsArray{<:NamedDimsArray}` to return/use names? That seems like the alternative option to me. The issue with that approach is that there might be a bunch of methods necessary for dealing with `NamedDimsArray` × `PermutedDimsArray` interactions. Could be prohibitively troublesome.

Reply:
Couldn't you argue the same for `Transpose`? The only difference is that the function usually called is named `transpose`.

Reply:
Ah right, that might mean that we can just include `PermutedDimsArray` with existing transpose-handling code (with some augmentation).

Reply:
Which functions do you have in mind? I just discovered that `*` does seem to know about this, `PermutedDimsArray(rand(2,3),(2,1)) * rand(2)` works... but does that cause any surprises?

Reply:
According to @oxinabox this would be a bunch of work so I'll defer to him

Reply:
I dislike this behaviour

Reply:
I think Eric's suggestion is nice, but is too much work and maintainence for the gain

Reply:
To be clear, I think this is an interesting case... but I find `PermutedDimsArray(::NamedDimsArray, perm) -> NamedDimsArray{PermuteDimsArray}` to be really surprising behaviour. I expect to get a `PermuteDimsArray{NamedDimsArrays}` which is what we get on master. I _do_ think allowing `perm` to be names i.e `Symbol`s make sense.

I'd be happy with special behaviour for `PermuteDimsArray{NamedDimsArrays}` if the maintainence cost is propornate.

Just a personal opinion - open to alternative views :)

Reply:
It does seem odd that Julia has `transpose` which makes `Transpose` but no `lazypermutedims` which makes `PermutedDimsArray`, so the name of the type does double duty. Maybe that's the core here? This PR wants to extend `lazypermutedims`, although it happens not to have that spelling. 

It would seem very strange to me that functions like `sum(A, dims=:k)` would be expected to work on things not of the form `NamedDimsArray{Storage}`. Anyone writing a function should then have to know to look inside, and not just dispatch on `::NamedDimsArray`. They don't have to do this with `NamedDimsArray{Transpose}` etc.

Reply:
> It would seem very strange to me that functions like sum(A, dims=:k) would be expected to work on things not of the form NamedDimsArray{Storage}

For the case of wrapper arrays, then unless they are also messing with `dims`, I expect them to delegate that kwarg straight to the inner array type. And if they are, I kind of expect in many cases (Inc `PermuteDimsArray`) they can do their messing about with it and passes them modified version to the wrapped array.
And so it just works,in theory.

Orthogonal features added via wrapper arrays should work no matter the order of the wrapping

Reply:
> Orthogonal features added via wrapper arrays should work no matter the order of the wrapping

OK, I like this vision of commutative wrappers. But for instance Tracker really needs to be outermost, I think. (I haven't looked at how you made this work.) Maybe it's a hard problem to decide who wins. 

Some functions now seem to dispatch on `::NamedDimsArray`. Perhaps this could be solved by some clever `Union{NDA, Adjoint{NDA}, Transpose{NDA}, ...}`? And some Base functions fail to pass things they see as illegal along, I guess you could also give them extra methods?
```julia
julia> xy = NamedDimsArray{(:x, :y)}(rand(2,2)); 
julia> yx = NamedDimsArray{(:y, :x)}(rand(2,2));

julia> xy * transpose(yx) # correctly errors
ERROR: DimensionMismatch("Cannot take matrix product of arrays with different inner dimension names. (:x, :y) vs (:x, :y)")

julia> xy * PermutedDimsArray(yx, (2,1)) # silently falls back to AbstractArray method
2×2 NamedDimsArray{(:x, :_),Float64,2,Array{Float64,2}}:
 0.571736  0.94319 
 0.379215  0.585958

julia> sum(PermutedDimsArray(yx, (2,1)), dims=:y) # doesn't pass :y through
ERROR: MethodError: no method matching iterate(::Symbol)
```

Reply:
(I seem to have deleted my comment by accident, sorry)

> Orthogonal features added via wrapper arrays should work no matter the order of the wrapping

I like this vision, but for now it seems many things do care. Some examples:

```julia
julia> xy = NamedDimsArray{(:x, :y)}(rand(2,2)); 

julia> yx = NamedDimsArray{(:y, :x)}(rand(2,2));

julia> xy * transpose(yx) # correctly errors, as NamedDimsArray is outermost
ERROR: DimensionMismatch("Cannot take matrix product of arrays with different inner dimension names. (:x, :y) vs (:x, :y)")

julia> xy * PermutedDimsArray(yx, (2,1)) # silently falls back to AbstractArray
2×2 NamedDimsArray{(:x, :_),Float64,2,Array{Float64,2}}:
 0.571736  0.94319 
 0.379215  0.585958

julia> sum(PermutedDimsArray(yx, (2,1)), dims=:y) # doesn't know how to pass this along
ERROR: MethodError: no method matching iterate(::Symbol)
```
Perhaps one could avoid these by always dispatching on some big `Union{NDA, Adjoint{NDA}, Transpose{NDA}, ...}` type, I don't know how messy that would be. 

But without commutativity of wrapping, I guess deciding who goes outermost might be a hard problem. I suppose my implicit rule here was that `PermutedDimsArray` is just a storage detail, like `Transpose`. User code shouldn't care, so it doesn't need to be exposed.